### PR TITLE
[Kernel] Fuse temperature + softmax in sampling for decode speedup

### PR DESCRIPTION
**Pull request by Godmook**
## Motivation

The standard sampling path in `sampler.py` executes temperature scaling and softmax as two separate CUDA kernels:

```python
logits.div_(sampling_info.temperatures)       # Kernel 1
logits[:] = torch.softmax(logits, dim=-1)     # Kernel 2
```

This is called on every decode step and incurs:
- **2 kernel launches** (~10us overhead each)
- **6 global memory passes** over the full `[batch_size, vocab_size]` tensor (4 reads + 2 writes)

For large-vocab models (Llama 3 128K, Qwen 152K), this becomes a meaningful bottleneck in the decode latency budget.

## Modifications

**New file: `python/sglang/srt/layers/fused_sampling.py`**

Triton kernel that fuses `div_(temperature) + softmax` into a single kernel with two variants:

- **Single-pass kernel** (vocab ≤ 32768): Loads entire vocab into registers once. Computes max, exp, sum, normalize all in-register. Only **1 read + 1 write = 2 memory passes**.
- **Multi-pass kernel** (vocab > 32768): 2-pass online softmax with `@triton.autotune` over (BLOCK_SIZE, num_warps, num_stages). **2 reads + 1 write = 3 memory passes**.

Both variants apply `logits / temperature` during the load, avoiding a separate division pass.

**Modified: `python/sglang/srt/layers/sampler.py`** (3 lines changed)

- Import `fused_temperature_softmax_inplace`
- Replace the `div_ + softmax` two-line sequence with the fused kernel in the standard sampling path
- Original implementation kept as fallback for non-CUDA platforms

**New file: `test/registered/sampling/test_fused_temperature_softmax.py`**

14 correctness tests comparing fused output against the reference (`div_ + softmax`) across:
- Batch sizes: 1, 2, 16, 64, 128, 512
- Vocab sizes: 1024, 32000, 128256
- Temperatures: 0.01 to 100.0
- Dtypes: bf16, fp16, fp32
- Edge cases: empty batch, temperature=1.0

**New file: `benchmark/kernels/bench_fused_temperature_softmax.py`**

Microbenchmark using `torch.cuda.Event` timing (200 iters, 50 warmup).

## Accuracy Tests

All 14 tests pass — fused output matches reference within tolerance (atol=1e-4, rtol=1e-3 for bf16):

```
test_fused_temperature_softmax.py::TestFusedTemperatureSoftmax::test_basic PASSED
test_fused_temperature_softmax.py::TestFusedTemperatureSoftmax::test_large_vocab PASSED
test_fused_temperature_softmax.py::TestFusedTemperatureSoftmax::test_batch_sizes PASSED
test_fused_temperature_softmax.py::TestFusedTemperatureSoftmax::test_temperature_one PASSED
test_fused_temperature_softmax.py::TestFusedTemperatureSoftmax::test_very_low_temperature PASSED
test_fused_temperature_softmax.py::TestFusedTemperatureSoftmax::test_very_high_temperature PASSED
test_fused_temperature_softmax.py::TestFusedTemperatureSoftmax::test_fp16_input PASSED
test_fused_temperature_softmax.py::TestFusedTemperatureSoftmax::test_fp32_input PASSED
test_fused_temperature_softmax.py::TestFusedTemperatureSoftmax::test_mixed_temperatures PASSED
test_fused_temperature_softmax.py::TestFusedTemperatureSoftmax::test_empty_batch PASSED
test_fused_temperature_softmax.py::TestFusedTemperatureSoftmax::test_inplace_basic PASSED
test_fused_temperature_softmax.py::TestFusedTemperatureSoftmax::test_inplace_bf16 PASSED
test_fused_temperature_softmax.py::TestFusedTemperatureSoftmax::test_inplace_large_vocab PASSED

13 passed
```

## Benchmarking and Profiling

Measured on NVIDIA A40 (SM 8.6), bf16, `torch.cuda.Event` timing, 200 iters after 50 warmup:

| bs | vocab | original (us) | fused inplace (us) | speedup |
|----|-------|--------------|-------------------|---------|
| 1 | 32000 | 25.5 | 23.4 | **1.09x** |
| 1 | 128256 | 43.0 | 33.8 | **1.27x** |
| 32 | 32000 | 28.5 | 21.3 | **1.34x** |
| 32 | 128256 | 150.5 | 46.7 | **3.22x** |
| 128 | 32000 | 140.9 | 34.7 | **4.06x** |
| 128 | 128256 | 612.3 | 177.1 | **3.46x** |
| 512 | 32000 | 541.7 | 121.4 | **4.46x** |
| 512 | 128256 | 2416.8 | 694.8 | **3.48x** |

**Speedup across all configurations: 1.09x–4.46x.** Improvement comes from:
- Eliminating 1 kernel launch (~10us)
- Reducing memory passes from 6 to 2 (single-pass) or 3 (multi-pass)
- Autotune selecting optimal (BLOCK_SIZE, num_warps, num_stages) per vocab size

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

### CODE REVIEW

File: python/sglang/srt/layers/sampler.py [hunk 1]
```diff
@@ -27,6 +27,15 @@
...
         top_p_renorm_prob,
     )
+
+_use_fused_sampling = False
+if is_cuda():
+    try:
+        from sglang.srt.layers.fused_sampling import fused_temperature_softmax_inplace
+
+        _use_fused_sampling = True
+    except ImportError:
```
Comment (DarkSharpness) [inline python/sglang/srt/layers/sampler.py]: Why except `ImportError` here? On cuda platform, triton always exists. There should be no import error here.

File: python/sglang/srt/layers/fused_sampling.py [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+    triton.Config({"BLOCK_SIZE": 16384}, num_warps=32),
+    triton.Config({"BLOCK_SIZE": 16384}, num_warps=32, num_stages=4),
+    triton.Config({"BLOCK_SIZE": 32768}, num_warps=32),
+    triton.Config({"BLOCK_SIZE": 32768}, num_warps=32, num_stages=4),
+]
+
+
+@triton.autotune(configs=_MULTI_PASS_AUTOTUNE_CONFIGS, key=["vocab_size"])
+@triton.jit
+def _multi_pass_temperature_softmax_kernel(
```
Comment (DarkSharpness) [inline python/sglang/srt/layers/fused_sampling.py:108]: FYI, have you compared the performance of the implementation with `flashinfer.sampling.softmax` which also supports passing an optional `temperature`?

**Review by DarkSharpness — CHANGES_REQUESTED** [review]

File: python/sglang/srt/layers/sampler.py (hunk 1 above)
Comment (Godmook) [reply python/sglang/srt/layers/sampler.py]: Nice Catch! I removed the try/except. Now it does a direct import under `if is_cuda()`.

File: python/sglang/srt/layers/fused_sampling.py (hunk 2 above)
Comment (Godmook) [reply python/sglang/srt/layers/fused_sampling.py:108]: Thanks for the pointer! I looked into this — `flashinfer.sampling` doesn't have a `softmax` function
with a temperature parameter. Temperature scaling and softmax are separate processors
(`flashinfer.logits_processor.Temperature` and `flashinfer.logits_processor.Softmax`).

FlashInfer's `LogitsPipe` can internally fuse them when compiled, but that requires adopting
the full `LogitsPipe` abstraction rather than a simple drop-in kernel call. Our Triton kernel
provides a lightweight alternative that fuses both into a single pass without requiring
the `LogitsPipe` framework.

Happy to add a comparison if there's a specific API you had in mind!

Comment (DarkSharpness) [reply python/sglang/srt/layers/fused_sampling.py:108]: try this: https://github.com/flashinfer-ai/flashinfer/blob/d426b1831e921c1226d059d19d09d433b4fece96/flashinfer/sampling.py#L722-L777

Comment (Godmook) [reply python/sglang/srt/layers/fused_sampling.py:108]: Thanks for the pointer! Benchmarked against `flashinfer.sampling.softmax` on A100.

### Results

| bs | vocab | triton_ip (μs) | flashinfer (μs) | speedup |
|----|-------|---------------|-----------------|---------|
| 1 | 32K | 63.0 | 61.5 | 0.98x |
| 1 | 128K | 77.2 | 62.0 | 0.80x |
| 32 | 32K | 61.3 | 64.2 | 1.05x |
| 32 | 128K | 77.1 | 86.1 | 1.12x |
| 128 | 32K | 62.5 | 74.1 | 1.19x |
| 128 | 128K | 118.8 | 308.2 | 2.59x |
| 512 | 32K | 110.8 | 259.5 | 2.34x |
| 512 | 128K | 463.6 | 1183.2 | 2.55x |

### Profiling (bs=512, vocab=128K)

Profiled `flashinfer.sampling.softmax` with `torch.profiler`:

| Name | CUDA time avg |
|------|---------------|
| OnlineSoftmaxFusedKernel | 680.5μs (53%) |
| unrolled_elementwise_kernel | 413.8μs (32%)  ← clone |
| Memcpy DtoD | 188.8μs (15%) |

No workspace allocation overhead — the flashinfer kernel itself takes ~680μs per call at this config. Correctness is identical between both implementations (`max diff: 4.66e-10`).

### Summary

- **bs ≤ 32**: flashinfer is slightly faster or comparable
- **bs ≥ 128**: Triton in-place kernel is **~1.9-2.6x faster**

Since continuous batching decode typically runs at higher batch sizes where throughput matters most, the Triton kernel provides a meaningful improvement in the common case.

@DarkSharpness

**Review by DarkSharpness — APPROVED** [review]
LGTM. I just wonder how long it would take to compile & autotune this triton kernel? Also, the autotune must be done in warm-up stage and should not introduce extra runtime overhead.

**Review by DarkSharpness — APPROVED** [review]

File: python/sglang/srt/layers/sampler.py [hunk 3]
```diff
@@ -27,6 +28,15 @@
         top_k_renorm_prob,
         top_p_renorm_prob,
     )
+
+    from sglang.srt.layers.fused_sampling import fused_temperature_softmax_inplace
```
Comment (BBuf) [inline python/sglang/srt/layers/sampler.py:32]: Can we move this import to the top of the file?

**Review by BBuf — APPROVED** [review]

File: python/sglang/srt/layers/sampler.py (hunk 3 above)
Comment (Godmook) [reply python/sglang/srt/layers/sampler.py:32]: fused_sampling pulls in Triton at import time. we keep this import inside if is_cuda(): with the other CUDA-only deps (flashinfer, sgl_kernel) so sampler stays importable on non-CUDA builds.

Thanks for help me :)
